src/contracts/strategy.ts

```
@@ -0,0 +1,102 @@
+import { BigNumber, Contract } from 'ethers';
+import { request, gql } from 'graphql-request';
+
+import { wallet } from '../providers';
+
+import configByNetwork from '../config';
+import { abi as anchorUSTStratABI } from '../abis/AnchorUSTStrategy';
+import { server } from '../api';
+
+type DepositOperation = {
+  id: string;
+  idx: string;
+  underlyingAmount: string;
+  ustAmount: string;
+};
+
+type RedeemOperation = {
+  id: string;
+  idx: string;
+  aUstAmount: string;
+};
+
+const config = configByNetwork();
+
+export const strategy: Contract = new Contract(
+  config.strategy,
+  anchorUSTStratABI,
+  wallet,
+);
+
+export async function finalizeDeposits() {
+  const operations = await depositOperations();
+
+  for (const op of operations) {
+    const operation: DepositOperation = op;
+
+    try {
+      const gasLimit = await strategy.estimateGas.finishDepositStable(
+        BigNumber.from(operation.idx),
+      );
+
+      await strategy.finishDepositStable(BigNumber.from(operation.idx), {
+        gasLimit,
+      });
+
+      break;
+    } catch (e) {
+      server.log.error((e as Error).message);
+    }
+  }
+}
+
+export async function finalizeRedemptions() {
+  const operations = await redeemOperations();
+
+  for (const op of operations) {
+    const operation: RedeemOperation = op;
+
+    const gasLimit = await strategy.estimateGas.finishRedeemStable(
+      BigNumber.from(operation.idx),
+    );
+
+    try {
+      await strategy.finishRedeemStable(BigNumber.from(operation.idx), {
+        gasLimit,
+      });
+
+      break;
+    } catch (e) {
+      server.log.error((e as Error).message);
+    }
```

I'm still confused by all of this, why do you need the try catch? Can't you just let it blow up?

---

we can't just let it blow up when we estimate the transaction because we use the success/failure scenarios of that call to decide what the consequent logic should be

if the estimation fails: operation not ready to be redeemed, try the next one
if the estimation succeeds: operation ready to be redeemed. redeem it and break the loop

if we don't handle the estimation failure bull will just retry over and over again and the other operations will never be finalized even if they can be

---

okok, I undestand, damn this is confusing. Can we make it clearer? Maybe extract some pieces to auxiliary functions or/and add some comments? I imagine myself reading this code in 3/4 months and not remembering why it's written like this

---

ok done!

I added an auxiliary function and comments to clarify why the `continue` and `break` statements happen